### PR TITLE
[FIX JENKINS-51495] descriptorRadioList does not honor DescriptorVisibilityFilter

### PR DESCRIPTION
**Pull request by matsushou**
See [JENKINS-51495](https://issues.jenkins-ci.org/browse/JENKINS-51495).

### Proposed changelog entries

Oleg's version:
* Update descriptorRadioList form elements to honor DescriptorVisibilityFilter extension points

Original version
*  Quit using index to associate descriptor selected on the front-end with the implementation on the back-end, and begin using descriptor id instead.

### Submitter checklist

- [ ] JIRA issue is well described
- [ ] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### CODE REVIEW

**Comment by batmat** [comment]
Test failure seems real?

**Comment by matsushou** [comment]
Yes, it seems real, and all 5 failures are due to not updating DescriptorExtensionList...
I've added a similar commit.

**Comment by matsushou** [comment]
@batmat 

Could you review this PR again (or remove needs-fix label in order others can review) ?

**Review by oleg-nenashev — APPROVED** [review]
LGTM, thanks @matsushou !

**Review by fcojfernandez — APPROVED** [review]
LGTM

**Review by jvz — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
i will land it into the next weekly if there is no negative feedback

**Comment by daniel-beck** [comment]
Would like to see a test, because `DescriptorVisibilityFilter` is not something regularly encountered AFAICT.

**Comment by matsushou** [comment]
Jenkins failed because of the error below.

`Failed to install https://download.java.net/java/GA/jdk11/9/GPL/openjdk-11.0.2_windows-x64_bin.zip to C:\Jenkins\tools\hudson.model.JDK\jdk11`

For me, it seems to be an environmental problem, not sure though...
I guess re-run may work.

**Comment by oleg-nenashev** [comment]
https://issues.jenkins-ci.org/browse/INFRA-2086 FYI. The Oracle's service is still broken, not that much I can do about it without access to the infra.

putting on hold so that you can discuss @daniel-beck 's comment

**Comment by daniel-beck** [comment]
@matsushou Only affects the Windows PR builds, please ignore them for now.

**Comment by matsushou** [comment]
> Only affects the Windows PR builds, please ignore them for now.

Thank you for your comment.
At last, is this PR assumed to be ready-for-merge again?

**Comment by oleg-nenashev** [comment]
IIUC https://github.com/jenkinsci/jenkins/pull/3969#issuecomment-487283884 is still open @matsushou

**Comment by matsushou** [comment]
sorry, I misunderstood the meaning of the comment...
I'll add some test cases shortly.

**Comment by oleg-nenashev** [comment]
@matsushou please let us know if you need any assistance

**Comment by matsushou** [comment]
thank you so much, I'll take time to solve it this weekend at the latest.

**Comment by matsushou** [comment]
I added a test case but couldn't run it locally... (need to wait Jenkins build)
Also, I'm not sure it's enough or an appropriate way to write a test in this case...

**Comment by matsushou** [comment]
began to fail without additional push... ❓ 

https://ci.jenkins.io/blue/organizations/jenkins/Core%2Fjenkins/activity?branch=PR-3969

**Review by oleg-nenashev — COMMENTED** [review]
Thatks, and sorry for the review delay. I checked the code, and I am not sure that returning null from `newInstanceFromRadioList` is actually an expected behavior. Maybe it makes sense to just throw an error. Will double-check the code to provide a definite answer

**Comment by oleg-nenashev** [comment]
Still om my side :(

**Comment by daniel-beck** [comment]
There is another use case for this:

When installation of a plugin moves the indexes around. If the config form was loaded before the plugin's extensions were loaded, but submitted after, the index in the list will not match the type, and an error will be shown to the user.

**Comment by MRamonLeon** [comment]
@oleg-nenashev how can we move it on? Are you going to look at it and provide an answer? thanks!

**Comment by oleg-nenashev** [comment]
Sorry, I missed the comments. Please request m reviews when needed so that it appears on my dashboard

**Review by oleg-nenashev — COMMENTED** [review]
I will submit a follow-up patch for it. Creating a null object creates a lot of potential problems down the line in usages, including unintentional security settings disabling if the property is filtered.

I will create a follow-up the request to suggest my version

**Comment by alecharp** [comment]
Closing to trigger a new build on Jenkins.

**Comment by timja** [comment]
> I will submit a follow-up patch for it. Creating a null object creates a lot of potential problems down the line in usages, including unintentional security settings disabling if the property is filtered.
> 
> I will create a follow-up the request to suggest my version

to be clear are you blocking this @oleg-nenashev? or happy for it to be merged?

**Comment by oleg-nenashev** [comment]
Sorry, I forgot to act on this pull request. In my TODO list for this week

**Review by oleg-nenashev — APPROVED** [review]
I reviewed the code, and I confirm that the suggested change is safe.

* All usages: https://github.com/search?q=org%3Ajenkinsci+newInstanceFromRadioList&type=Code
* There are 3 usages in the Jenkins Core. All of them handle nulls properly. I will be submitting a follow-up pull request(s) with annotation fixes and one unrelated bug I discovered
* There is one usage in Groovy Plugin. https://github.com/jenkinsci/groovy-plugin/blob/6e423e98dd819baedbc15ec93410b1719e4cf574/src/main/java/hudson/plugins/groovy/AbstractGroovy.java#L35-L66 .  Metod - getScriptSource()
  * This usage is in the deprecated method, but apparently it is used in the Configuration Slicing plugin: https://github.com/jenkinsci/configurationslicing-plugin/blob/2fa473ed02142b0e97341f9b3b182ea8ac4c52cc/src/main/java/configurationslicing/tools/GroovySlicer.java#L46 . This is the functionality available to admins only
  * Configuration Slicer Plugin invokes Groovy execution. When it is ran, https://github.com/jenkinsci/groovy-plugin/blob/6e423e98dd819baedbc15ec93410b1719e4cf574/src/main/java/hudson/plugins/groovy/Groovy.java#L64-L67 will process `null` and fail the execution when the required descriptor is filtered by DescriptorVisibilityFilter 
  * Although this behavior is not ideal, IMO it does not constitute to a horrible regression. Admins will be able to track down the problem if needed, and I would argue that it is better to fail the execution if it is filtered on the instance
* There is one usage in the proprietary plugins I have access to. This usage handles nulls properly

**Review by res0nance — COMMENTED** [review]
Looks ok, https://github.com/jenkinsci/jenkins/pull/3969#issuecomment-487283884 still seems open though.

File: test/src/test/java/hudson/ExtensionListTest.java [hunk 1]
```diff
@@ -174,6 +177,28 @@ public void legacyDescriptorList() throws Exception {
         assertEquals(0,LIST.size());
     }
 
+	@Test
+	public void newInstanceFromRadioList() throws Exception {
+		// test for DescriptorList
```
[the remark this answers is not in the document]
Comment (oleg-nenashev) [reply test/src/test/java/hudson/ExtensionListTest.java]: Fixed, thanks!

**Comment by oleg-nenashev** [comment]
I plan to merge it in 24 hours if no negative feedback

**Comment by oleg-nenashev** [comment]
Thanks a lot @matsushou ! Sorry that it took so much time to get it merged

**Pull request merged**